### PR TITLE
Semi-additive (Support alternate granularity)

### PR DESCRIPTION
**Pull request by WilliamDee**
## Extension of https://github.com/transform-data/metricflow/pull/183
Previously, we supported semi-additive measures in a way that you would only be allowed to take the non-additive dimension and perform a MIN/MAX aggregation against the full time range and get values from that big "window". In this PR, we break this down such that if the `non_additive_dimension` is passed in the query as a group by, it would do a `GROUP BY` in the inner query which enables us to have `windows` of the granularity passed in the query.

These are the two scenarios 
- `query --metrics ... --dimensions non_additive_dim(__week/__month/__year)`
    - Uses the granularity of `non_additive_dim` as the window to perform min/max
- `query --metrics ... --dimensions dim1 (--where non_additive_dim between X AND Y)`
    - Does a min/max of the full window (either `FULL RANGE` or between `X and Y`)

## Changes
- defaulted `window_choice` to `MIN` instead of required
- Added the ability to group by a TimeDimension in the `SemiAdditiveJoinNode` inner query so that we can perform window filters on different granularity windows instead of just on the full time range.
- Refactored the way we build the `SemiAdditiveJoinNode`
  - Reverted the changes in `SourceNodeBuilder`
  - Build the `SemiAdditiveJoinNode` in the `DataflowPlanBuilder` so that we can alter the node depending on the query spec
  - This also fixes the bug in which constraints was not working properly for semi additive measures

## Examples
```
if we have a data set that includes "account_balances,user,date", we can build a
"latest_account_balance_by_user" data set using this node by filtering the data set such that for each user,
we get the MAX(date) and return the account_balance corresponding to that date.

Data transformation example,
| date       | account_balance | user |                             | date       | account_balance | user |
|:-----------|-----------------|-----:|     identifier_specs:       |:-----------|-----------------|-----:|
| 2019-12-31 |            1000 |    u1|       - user                | 2020-01-03 |            2000 |    u1|
| 2020-01-03 |            2000 |    u1| ->  time_dimension_spec: -> | 2020-01-12              1500 |    u2|
| 2020-01-09 |            3000 |    u2|       - date                | 2020-01-12 |            1000 |    u3|
| 2020-01-12 |            1500 |    u2|     agg_by_function:
| 2020-01-12 |            1000 |    u3|       - MAX


Similarly, if we don't provide any identifier_specs, it would end up performing the aggregation filter only on the
time_dimension_spec without any grouping by any identifiers.

Data transformation example,
| date       | account_balance | user |                             | date       | account_balance |
|:-----------|-----------------|-----:|     identifier_specs:       |:-----------|----------------:|
| 2019-12-31 |            1000 |    u1|                             | 2020-01-12 |            2500 |
| 2020-01-03 |            2000 |    u1| ->  time_dimension_spec: ->
| 2020-01-09 |            3000 |    u2|       - date
| 2020-01-12 |            1500 |    u2|     agg_by_function:
| 2020-01-12 |            1000 |    u3|       - MAX


Additionally, we can aggregate against 'windows' of a dataset. For example, if we group by a non_additive time dimension,
we perform the MIN/MAX filtering on the granularity window provided in that group by instead of the full time range.

Data transformation example,
| date       | account_balance | user |                                  | date       | account_balance |
|:-----------|-----------------|-----:|     identifier_specs:            |:-----------|----------------:|
| 2019-12-31 |            1500 |    u1|     time_dimension_spec:         | 2019-12-31 |            1500 |
| 2020-01-03 |            2000 |    u1| ->    - date                  -> | 2020-01-07 |            3000 |
| 2020-01-09 |            3000 |    u2|     agg_by_function:             | 2020-01-14 |            3250 |
| 2020-01-12 |            1500 |    u2|       - MIN
| 2020-01-14 |            1250 |    u3|     queried_time_dimension_spec:
| 2020-01-14 |            2000 |    u2|       - date__week
| 2020-01-15 |            4000 |    u1|
```

### CODE REVIEW

File: metricflow/dataflow/builder/dataflow_plan_builder.py
Comment (tlento) [inline metricflow/dataflow/builder/dataflow_plan_builder.py:465]: These two accessors are making assumptions about how this method will be called, and they'll produce incorrect results if those assumptions ever change.

I'd prefer to make this part of the method signature, or find some other way of ensuring that this is correct. 

We could also bundle these up top, where the assertion about the data sources lives, and add invariant checks at that point.

File: metricflow/dataflow/builder/dataflow_plan_builder.py
Comment (tlento) [inline metricflow/dataflow/builder/dataflow_plan_builder.py:471]: Does this work with custom measure agg time dimensions? The `find_primary_time_dimension` checks at the data source level, and I don't see anything looking at the measure itself, but the measure might well have a different time dimension set. Does the source node builder step deal with this for us? If so it'd be good to document this behavior, and I recommend doing it via a helper method that returns the queried_time_dimension_spec and can then be documented independently with its assumptions.

It's actually kind of unclear to me how all of these things fit together, honestly, so more clarity in this block would be super helpful.

File: metricflow/model/semantics/semantic_containers.py
Comment (tlento) [inline metricflow/model/semantics/semantic_containers.py:254]: See this goes to the DataSource, not the source node set, which is why I was wondering if this accounts for metric_time usage on measures with a different agg_time_dimension.

It's possible for a data source to have no primary time dimension, I think, or it used to be. Regardless it's possible for data_source.primary_time to never be referenced anywhere.

**Review by tlento — COMMENTED** [review]
Looks super promising! I have some questions and minor comments inline. I think it'd be good if @plypaul took a look as well, he might see some things we're missing.

File: metricflow/model/objects/elements/measure.py
Comment (WilliamDee) [reply metricflow/model/objects/elements/measure.py:19]: I talked with Nick here, we wanted to default to MIN simply because of how our granularity works and the preference of a minimal config. For example, if we did `ds__week` and say we get 3 rows of `2022-01-01`, `2022-01-08`, `2022-01-15` if we had a MIN default, the result would look like the following,
```
2022-01-01  | value at 2022-01-01
2022-01-08  | value at 2022-01-08
2022-01-15  | value at 2022-01-15
```
If we did max, the result would look like this 
```
2022-01-01  | value at 2022-01-07
2022-01-08  | value at 2022-01-14
2022-01-15  | value at 2022-01-21
```
so intuitively it makes more sense to default as MIN rather than MAX. Open for further discussion here though or we could just simply keep this as required until we decide which is better.


As for the `AggregationType`, I agree with it being weird, but I was going to add a validation on that to ensure it's `MIN/MAX`.  I was thinking that would be enough and makes sure it gets blocked on usage of other types.

File: metricflow/plan_conversion/dataflow_to_sql.py
Comment (WilliamDee) [reply metricflow/plan_conversion/dataflow_to_sql.py:1464]: yea, we don't actually access this anywhere outside the subquery. I made this change since I actually had issues with BQ where the subquery was something like this 
```sql
SELECT 
  MIN(ds) AS ds
FROM data_set
GROUP BY ds
```
this threw a `cannot group by an aggregate` error. So by making this local alias, it allowed it to group by properly. So unless we might have a use case of `MIN(ds)` anywhere outside of the subquery I assumed it would be fine to do this.

File: metricflow/model/objects/elements/measure.py
Comment (tlento) [reply metricflow/model/objects/elements/measure.py:19]: The validation is fine for now, but a separate enum helps us out more, I think, because we can rely on the typechecker (and then the validation comes along with the JSON schema update).

The MIN thing makes sense, it's worth documenting that reasoning somewhere. I think the user-facing docs are the right place, actually, as long as it's not too in the weeds.

File: metricflow/model/semantics/semantic_containers.py
Comment (WilliamDee) [reply metricflow/model/semantics/semantic_containers.py]: now that I think about it, you're right. This doesn't cover the case of measures with different agg_time_dimension. With more thought on that, I actually realized we didn't need to do any of the finding primary time dimension stuff. Since the source node splits the nodes into measures with the same agg_time_dimension. I can just track the measure's agg_time_dimension and compare that to the non_additive_dimension_spec to check whether `metric_time` is a match or not.  So I ended up wiping all the code in favour of that. https://github.com/transform-data/metricflow/pull/253/commits/da5ce02dc2b0922e3a90f505e450c9abeea8f24b

File: metricflow/dataflow/builder/dataflow_plan_builder.py
Comment (WilliamDee) [reply metricflow/dataflow/builder/dataflow_plan_builder.py]: addressed here https://github.com/transform-data/metricflow/pull/253#discussion_r977898952

File: metricflow/dataflow/builder/dataflow_plan_builder.py
Comment (WilliamDee) [reply metricflow/dataflow/builder/dataflow_plan_builder.py]: addressed here https://github.com/transform-data/metricflow/pull/253#discussion_r977898952

File: metricflow/dataflow/dataflow_plan.py
Comment (tlento) [reply metricflow/dataflow/dataflow_plan.py:487]: Never mind, I think I was getting confused by the identifier in the data itself.

The contract for semi-additive measures is that we can aggregate *within* values for the non-additive dimension, but not *across* them, and this is exactly what happens here. While it might be odd for somebody to ignore the `user` entry in this case, it's not wrong - in fact, if they really don't care about the `user` under normal circumstances (or if they're never allowed to have user-level aggregations in their reporting pipeline) this is exactly what they'd want.

File: metricflow/model/parsing/schemas.py
Comment (tlento) [inline metricflow/model/parsing/schemas.py:158]: For validating aggregation type we could consider making it a restricted enum in the JSON schema here. It does seem cleaner to add a custom enum type for this, though - I suspect we'll need it for stuff like SCDs as well.

File: metricflow/dataflow/builder/dataflow_plan_builder.py
Comment (tlento) [inline metricflow/dataflow/builder/dataflow_plan_builder.py:321]: Nit - shorten this to avoid the 120 char overrun.

```
"""Finds the TimeDimensionSpec matching the non_additive_dimension_spec, if any"""
```

File: metricflow/dataflow/builder/dataflow_plan_builder.py
Comment (tlento) [inline metricflow/dataflow/builder/dataflow_plan_builder.py:315]: This is a very generic name. Do we want to make it more non-additive specific?

File: metricflow/dataflow/builder/dataflow_plan_builder.py
Comment (tlento) [inline metricflow/dataflow/builder/dataflow_plan_builder.py:487]: These are still unchecked. If you look on line 349 we have an invariant check enforcing the internal contract for `_find_measure_recipe`. I think we need to extend that check, because we are now relying on this extended set of conditions (i.e., there is only one agg time dimension and only one non-additive spec) internally.

You can update the contract in one of two ways - either do it via a runtime assertion, or update the signature to accept a non-additive dimension spec and an agg time dimension. The former is simplest and probably cleanest, especially given that this is an internal helper method. The latter is more ergonomic as the signature itself tells developers what to expect.

The best thing might be to overhaul the signature to take in an object representing the measure specs, with properties for data source, agg time dimension, and non-additive dimension. That object can then maintain the invariants and encapsulate the work of doing the extraction of these values at the callsites. But it's more work.

**Review by tlento — COMMENTED** [review]
I think this is good, but I'd value another set of eyes. Still not quite there, myself. Stupid COVID.

Anyway, please do add the invariant or helper object or whatever you deem most appropriate for ensuring `_find_measure_recipe` always gets appropriate input.

File: metricflow/plan_conversion/dataflow_to_sql.py
Comment (plypaul) [reply metricflow/plan_conversion/dataflow_to_sql.py:1464]: One of the invariants is that all column aliases are generated through the column association resolver. Is this something you can handle there?

Comment (WilliamDee) [reply metricflow/plan_conversion/dataflow_to_sql.py]: as discussed offline, we are adding `Optional[AggregationState]` to `resolve_time_dimension_spec` and just in this semi_additive_join scenario, I’ll be passing in `AggregationState.COMPLETE` and we would render that column alias as `{time_dim}__complete` and that suffix `*__complete` may get translated to something else that’s simpler in the future.

File: metricflow/dataflow/builder/dataflow_plan_builder.py
Comment (WilliamDee) [reply metricflow/dataflow/builder/dataflow_plan_builder.py]: I added the invariant checks with a helper function that takes in a `Sequence[MeasureSpec]` and it asserts that all those measure_specs has the same `agg_time_dimension` and `non_additive_dimension_spec`

File: metricflow/dataflow/builder/dataflow_plan_builder.py
Comment (plypaul) [inline metricflow/dataflow/builder/dataflow_plan_builder.py:337]: Are there issues if the user queries for `metric_time__week` and `metric_time__month`? Or is this validated somewhere?

Comment (WilliamDee) [reply metricflow/dataflow/builder/dataflow_plan_builder.py:337]: They would be allowed to do that

File: metricflow/dataflow/builder/dataflow_plan_builder.py
Comment (tlento) [inline metricflow/dataflow/builder/dataflow_plan_builder.py:509]: This is nice! I would love this to be part of the method signature, maybe containing the measure spec sequence as well. So this method would take in a MeasureSpecPropeties or MeasureSpecsByCommonProperties or whatever we end up calling it, and it's obvious to all of us that we want a collected set of measures. Maybe we could even put in the data source name and collect these at the very top, and then remove all of the validation checks inside of this method.

This is getting into separate PR territory, though, so I think it's fine as-is. Maybe add a TODO comment if you don't pull this into the method signature straightaway.

**Review by tlento — COMMENTED** [review]
Ugh time granularities. Do we need another validation? Might be worth a quick chat separately just to make sure we're doing the right things here.

As long as our granularity handling is correct here this seems good to go to me. If we need another validation to prevent weirdness we can add it to the validation PR.

File: metricflow/dataflow/builder/dataflow_plan_builder.py
Comment (WilliamDee) [reply metricflow/dataflow/builder/dataflow_plan_builder.py]: Did this in this PR!

File: metricflow/dataflow/builder/dataflow_plan_builder.py
Comment (WilliamDee) [reply metricflow/dataflow/builder/dataflow_plan_builder.py:339]: We are validating that in here https://github.com/transform-data/metricflow/pull/259

File: metricflow/test/integration/test_cases/itest_semi_additive_measure.yaml
Comment (tlento) [inline metricflow/test/integration/test_cases/itest_semi_additive_measure.yaml:227]: Newline! 😝

File: metricflow/test/model/validations/test_element_const.py
Comment (tlento) [inline metricflow/test/model/validations/test_element_const.py:40]: Thanks!

**Review by tlento — COMMENTED** [review]
Seems reasonable! We need to refactor the dataflow plan builder, it's getting out of control, but I think that's beyond the scope of this PR.

I just realized we should probably run a whole lot of tests so I'll leave this in reviewer queues for a bit in case @plypaul wants to weigh in.

File: metricflow/specs.py
Comment (WilliamDee) [reply metricflow/specs.py:312]: I was actually not aware of that! This shouldn't be needed then!

Comment (WilliamDee) [reply metricflow/specs.py:312]: ooo wait, just tested it, it doesn't actually work. It converts the values of the as a tuple in the order of the fields so it has nothing to do with the value itself. ie., if we had this
```
NonAdditiveDimensionSpec(
   name="asd", window_choice="min",window_groupings=(1,2)
)
```
it would convert it to 
```
("asd", "min", (1,2))
```
which would not be equal to say 
```
("asd", "min", (2,1))
```
which it should be equal

**Review by tlento — APPROVED** [review]
336 files changed?! We might need to revisit this whole snapshot business... 🗄️

**Comment by WilliamDee** [comment]
> 336 files changed?! We might need to revisit this whole snapshot business... 🗄️ 

Big change due to reverting the source node building logic which happens for every query!

**Pull request merged**
